src/Receipts/ReceiptResponse.php

```diff
@@ -1,90 +1,110 @@
 
     /**
+     * An indicator that an error occurred during the request.
      * @var bool|null
      */
     protected $isRetryable;
 
     /**
+     * The latest Base64 encoded app receipt.
+     * Only returned for receipts that contain auto-renewable subscriptions.
      * @var string|null
      */
     protected $latestReceipt;
 
     /**
-     * @var array|ReceiptInfo[]
+     * An array that contains all in-app purchase transactions.
+     * @var array|LatestReceiptInfo[]|null
      */
     protected $latestReceiptInfo;
 
     /**
-     * @var array|PendingRenewal[]
+     * In the JSON file, an array where each element contains the pending renewal information
+     * for each auto-renewable subscription identified by the product_id.
+     * @var array|PendingRenewal[]|null
      */
     protected $pendingRenewalInfo;
 
     /**
-     * @var Receipt|null
+     * the receipt that was sent for verification.
+     * @var array|null
      */
     protected $receipt;
 
     /**
-     * @var Status
+     * Either 0 if the receipt is valid, or a status code if there is an error.
+     * @see https://developer.apple.com/documentation/appstorereceipts/status
+     * @var int
      */
     protected $status;
 
     /**
-     * ReceiptResponse constructor.
-     * TODO: replace public constructor usage with a static factory method
-     * @param array $attributes
-     * @throws InvalidReceiptException
+     * @var bool
      */
-    public function __construct(array $attributes)
-    {
-        if ($attributes['status'] !== 0) {
-            throw InvalidReceiptException::create($attributes['status']);
-        }
+    private $parsedLatestReceiptInfo;
 
-        $this->environment = $attributes['environment'];
-        $this->latestReceipt = $attributes['latest_receipt'] ?? null;
+    /**
+     * @var bool
+     */
+    private $parsedPendingRenewalInfo;
 
-        $this->latestReceiptInfo = [];
-        foreach ($attributes['latest_receipt_info'] ?? [] as $itemAttributes) {
-            $this->latestReceiptInfo[] = ReceiptInfo::fromArray($itemAttributes);
-        }
+    /**
+     * ReceiptResponse Constructor
+     */
+    public function __construct(int $status)
+    {
+        $this->status = $status;
+        $this->parsedLatestReceiptInfo = false;
+        $this->parsedPendingRenewalInfo = false;
+    }
 
-        $this->receipt = isset($attributes['receipt']) ? Receipt::fromArray($attributes['receipt']) : null;
-        $this->status = new Status($attributes['status']);
+    /**
+     * Static factory method
+     * @param array $body
+     * @return ReceiptResponse
+     */
+    public static function fromArray(array $body): self
+    {
+        $obj = new self($body['status']);
 
-        $this->pendingRenewalInfo = [];
-        foreach ($attributes['pending_renewal_info'] ?? [] as $item) {
-            $this->pendingRenewalInfo[] = PendingRenewal::fromArray($item);
-        }
+        $obj->environment = $body['environment'] ?? null;
```

Comment:
i am thinking if we should implement a fail-fast here in case of missing of essential values to avoid unintended side-effects — what do you think?

Reply:
The only required attribute here is the `status`, the Constructor will fail-fast in case it's missed.

All other attributes are optional, in some cases (which is not documented) Apple does not send one or more of other keys.